### DISCUSSION TITLE
How can I move the legend outside the ResponsiveContainer?

### DISCUSSION THREAD

**Discussion by michaelschufi** [recharts/recharts#4971] [Q&A] [answered]
Hi 👋 

I'm having trouble because the responsive container seems to take the legend into account in its width and height calculation. This makes the actual chart hide itself when less space is available.

**Screenshot: Enough space available**
![grafik](https://github.com/user-attachments/assets/e175192d-fbe0-41d0-87e3-89ed2d6a4b5c)

**Screenshot: Less space available**
![grafik](https://github.com/user-attachments/assets/85bd87fb-e6ef-4bc6-85d9-cce856825701)

Because of this behaviour, I tried to move the `<Legend>` outside the `<ResponsiveContainer>`. But this didn't work because the `<PieChart>` has to be inside the  `<ResponsiveContainer>` and the `<Legend>` needs to be inside the `<PieChart>`.

**How can I properly move the `<Legend>` outside the `<ResponsiveContainer>` so it can take up its own space and the `<Pie>` itself can be responsive**?

**Comment by ckifer** [comment]
Short answer is that you can't yet. 

We just merged a change into the 3.x branch to support user defined legend portals where you will be able to render the legend into whatever container you wish. But it'll have to wait until 3.0

See https://github.com/recharts/recharts/pull/4947

**Reply by michaelschufi** [reply]
Okay, thanks! 😃 
Is there a beta where we can already test this? Or is it still a bit too early?

**Reply by ckifer** [reply]
Still a bit too early! Lots of changes made but still quite a few to go

**Comment by MardanovaA** [comment]
Hello! Is there any solution now?

**Reply by ckifer** [reply]
Yes you can use portals which were released in 3.0

**Reply by MardanovaA** [reply, marked as answer]
Yes, thank you!
I understand than i can use it, I just didn't understand how to do it because I couldn't find any information in the documentation.
I'll write this for anyone who might have a similar problem. You can use the legend portal like this:

```
export const LegendPortal = () => {
    const [legendPortal, setLegendPortal] = useState<HTMLDivElement | null>(null);

    return (
      <div style={{ display: 'flex', width: '100%', height: '100%' }}>
        <div
          style={{ width: '100%', height: '100%' }}
          ref={node => {
            if (legendPortal == null && node != null) {
              setLegendPortal(node);
            }
          }}
        /> // legend will be here
        <div style={{ width: '100%', height: '100%' }}>
          <ResponsiveContainer width="100%" height={surfaceHeight}>
            <AreaChart
              data={data}
            >
              {/* The target component */}
              {legendPortal && (
                <Legend
                  {...args}
                  wrapperStyle={{ position: 'relative', top: 'none', left: 'none', width: 'inherit', marginTop: 20 }}
                  formatter={() => 'Recharts Downloads'}
                  portal={legendPortal}
                />
              )}

              {/* YAxis,  XAxis, etc. */}
            </AreaChart>
          </ResponsiveContainer>
        </div>
      </div>
    );
}
```

**Reply by ckifer** [reply]
There is a page in the storybook but not in the website. Sorry I would've linked it but was just replying on mobile

**Reply by ckifer** [reply]
https://main--63da8268a0da9970db6992aa.chromatic.com/?path=/story/api-component-legend--legend-portal

**Reply by DevinCLane** [reply]
hi there, was working on this myself and noticed the stackblitz example listed in the storybook link didn't have the `useState` import or the JSON data that the chart is trying to read. Fixed those two here:

https://stackblitz.com/edit/v56yvezj?file=src%2FExample.tsx

cheers!
